Granulate changes to perf

DanielShaulov opened the pull request:
This PR is open so it is easy to view all of our changes to perf.
Each commit has a single change with detailed explanations.
No need to merge, we are using versions compiled on this branch.

Commit:
perf probe: Add support for DW_OP_call_frame_cfa vars

Add support for probes on variables with DW_OP_call_frame_cfa
as the dwarf operation in the debug info.

Some compilers (specifically Golang compiler) output
DW_OP_call_frame_cfa instead of DW_OP_fbreg for variables
on the stack. If DW_OP_call_frame_cfa is the only expression
than it is the same as DW_OP_fbreg with an offset of zero.
In the case of the Golang compiler, DW_OP_call_frame_cfa may
be followed by DW_OP_consts, with a number and than DW_OP_plus.
This trio is the same as DW_OP_fbreg with the number from
DW_OP_consts as the offset.

With this change, probing on functions in Golang with variables works.

Signed-off-by: Daniel Shaulov <daniel.shaulov@granulate.io>

Commit:
perf: Change libdw libraries to support elfutils>=0.178

Up untill now, static compilation of perf was still trying to dynamically
load the backends for libebl. Starting with version 0.178 of elfutils,
libebl no longer needs to dynamicall load backends. This means that
perf compiled statically with elfutils>=0.178 is trully static.

In elfutils 0.178, libebl is also no longer a standalone library,
it is included as part of libdw, so there is no longer a need for -lebl.
Also - some of the backends required -lpthread, and since they are now
included in the static build, we also need to add -lpthread.

This change does not support older versions of elfutils.

Signed-off-by: Daniel Shaulov <daniel.shaulov@granulate.io>

Commit:
perf probe: Make perf-probe use symbols when debuginfo is incomplete

When a binary has debug symbols, but they are incomplete,
some files compiled with debug info and some didn't, perf-probe will
look for the debuginfo, and when it fails to find it will refuse to add the probe.
If the debug info is removed entirely perf-probe gracefully degrades to using symbols.
This commit adds the fallback to symbols for cases where dwarf exists but fails

The issue was found with a nodejs binary which ships with debug info only
for the '.c' files but no for '.cc' (c++) files, and trying to probe a c++ function.

Signed-off-by: Daniel Shaulov <daniel.shaulov@granulate.io>

Commit:
perf probe: Add uint8* as a supported string type

When trying to cast a probe variable to string, the type is checked
to be "(unsigned) char *". In Golang debug info, the strings are
ultimately "uint8 *", so the cast fails.

This commit adds the type to the list of types that can be casted
to string, and also prints the type when it is found to not match
one the types, for easier debugging of future issues.

Signed-off-by: Daniel Shaulov <daniel.shaulov@granulate.io>

Commit:
perf: Add Dockerfile

Add a Dockerfile that can be used to build a static perf
The image is based on ubunut:16.04 to have a libc that supports older
kernel versions.

To build a static perf using the Dockerfile you can run the following (at the repo root):
docker build -t buildbox - < tools/perf/Dockerfile
docker run -v $PWD:/linux -w /linux/tools/perf buildbox sh -c "make clean all LDFLAGS=-static"

Signed-off-by: Daniel Shaulov <daniel.shaulov@granulate.io>

Jongy committed:
perf inject: Keep previous cmdline

Force "perf inject" to write the cmdline it has read from the input file,
instead of writing its own cmdline. "perf inject" cmdline is hardly useful,
"perf record"s is what we care about.

Do note it's not exactly complete - I did not change write_cmdline() so
that one still writes readlink("/proc/self/exe") of "perf inject" as the
argv[0], and we drop the argv[0] we've read from the file. Unless you use
different "perf" binaries for "record" and "inject", this doesn't matter.

Signed-off-by: Yonatan Goldschmidt <yonatan.goldschmidt@granulate.io>

Commit:
perf: Use exec_comm instead of comm when printing

Process names are much clearer as the first line in a flamegraph
If we ever want to upstream this, it needs to be a separate option

Jongy commented:
Rebased from https://github.com/Granulate/linux/commit/98a07a10b043b503d034e09af518f6985e97b2d8 (based on `v5.6`) to `v5.12`. Removed commits that were merged to mainline & commits that we've added+reverted.

`git range-diff` on my box:
```
$ git range-diff origin/master..v5.6g3-perf origin/master..HEAD
 1:  d959be3077e1 =  1:  35d307ea9c3a perf probe: Add support for DW_OP_call_frame_cfa vars
 2:  ba1792048a5c =  2:  3bfa53da1b73 perf: Change libdw libraries to support elfutils>=0.178
 3:  ce37bc52dd6f <  -:  ------------ perf inject --jit: Remove //anon mmap events
 4:  56994bef52d9 =  3:  2c7eac4062d4 perf probe: Make perf-probe use symbols when debuginfo is incomplete
 5:  b21ba3516dd1 =  4:  9be7a5ece056 perf probe: Add uint8* as a supported string type
 6:  577ab9572fb0 =  5:  ba2776008fb9 perf: Add Dockerfile
 7:  2b6b2682cd32 <  -:  ------------ perf inject: Fix header after injecting
 8:  646009c591c1 <  -:  ------------ perf tools: Add 'in_pidns' to nsinfo struct
 9:  1f0778a993bf <  -:  ------------ perf inject jit: Add namespaces support
10:  062f04b2db33 <  -:  ------------ Revert "perf inject: Fix header after injecting"
11:  c800a562814b =  6:  5ad1bebfc3ed perf inject: Keep previous cmdline
12:  8c3e6c8aff99 <  -:  ------------ perf: Use exec_comm instead of comm when printing
13:  98a07a10b043 <  -:  ------------ Revert "perf: Use exec_comm instead of comm when printing"
```

marcin-ol committed:
Take exec comm str to set child thread's comm when handling thread__fork event (#2)

Commit:
Revert "perf inject: Keep previous cmdline"

This reverts commit 5ad1bebfc3ed2d1255f11986c627d77a15912710.

Commit:
Merge tag 'v6.0' into granulate-perf
Linux 6.0

Jongy committed:
perf record: Allow using switch-output=time together with signal

The use case is, I'd like to switch upon signal but still protect with a timeout.

Commit:
Merge v6.7 to granulate-perf

slicklash committed:
perf: Add -F +symline option (#3)
